### PR TITLE
Let web UI requests to alertmanager be served by the distributor

### LINKED ISSUE #5090: [Alertmanager] Increasing number of AM replicas above the replication factor breaks AM
**Describe the bug**
When using alertmanager with sharding enabled, if we increase the number of AM instances above the replication factor, the UI breaks and sometimes return 'the Alertmanager is not configured' error. (See video)
When curling directly into the alertmanager pods using /alertmanager/api/v1/alerts, we also see that some pods return 'the Alertmanager is not configured' while others return correct responses and that tenants alerts.

https://user-images.githubusercontent.com/89514357/211770150-7e0b15c8-cb1c-4fa4-99a5-eb26f4943699.mp4

But if we decrease the number of instances to below the replication factor, it will return to normal and work okay.

For example

#works okay
replication_factor = 3
number of replicas <= 3

#starts breaking again
replication_factor = 3
number of replicas >= 4

**To Reproduce**
1. Start Cortex with just alertmanager in kubernetes using statefulset and below config
2. cortex.yml
```
alertmanager:
  enable_api: true
  external_url: /alertmanager
  sharding_enabled: true 
  sharding_ring:
    kvstore:
      store: consul 
      prefix: alertmanagers/
      consul:
        host: ${CONSUL_HOST}
        http_client_timeout: 20s
    replication_factor: 3
  alertmanager_client:
    grpc_compression: gzip
alertmanager_storage:
  backend: gcs
  gcs:
    bucket_name: ${BUCKET_NAME}
    service_account: ${SERVICE_ACCOUNT}
```
3. set replicas to 3. confirm alertmanager is working
4. set replicas to 4+. Alertmanager UI will break and some pods will return 'the Alertmanager is not configured' even if that tenant has been configured
**Expected behavior**
We should be able to set the replicas number to greater than  the replica_factor, without alertmanager breaking.

**Environment:**
 - Infrastructure: Kubernetes
 - Deployment tool: kustomize
 - kvstore: consul
 - alertmanager_storage: gcs

**Additional Context**
We see these logs when alertmanager starts breaking
`component=MultiTenantAlertmanager msg="deleted local directory for user" dir=data/test-user user=test-user`

### PR DESCRIPTION
**Pull request by ShunjiTakano**
Signed-off-by: Shunji Takano <shunji.takano.dev@gmail.com>

**What this PR does**:
Fixes alertmanager web UI requests and routes all requests to the Alertmanager Distributor component when sharding is enabled. Currently web ui requests are not served to the distributor.

**Which issue(s) this PR fixes**:
Fixes #5090

**Checklist**
- [ ] Tests updated *Not too sure how/where to go about new test cases. Would be glad to add some if I could get some pointers.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by friedrichg** [comment]
Thanks for your contribution!

I understand this probably makes the workflow simpler but at the same time will make an additional request for web UIs, probably increasing latency and maybe failing simple requests that could work.

Can you elaborate more why is this important?

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
 * [CHANGE] Alertmanager: Local file disclosure vulnerability in OpsGenie configuration has been fixed. #5045
 * [CHANGE] Rename oltp_endpoint to otlp_endpoint to match opentelemetry spec and lib name. #5067
 * [CHANGE] Distributor/Ingester: Log warn level on push requests when they have status code 4xx. Do not log if status is 429. #5103
+* [CHANGE] Alertmanager: Route web-ui requests to the alertmanager distributor when sharding is enabled. #5120
```
Comment (friedrichg) [inline CHANGELOG.md:8]: It's a more a bugfix than a change.

**Review by friedrichg — CHANGES_REQUESTED** [review]
Test cases are needed, at least add a test for ServeHTTP with multitenant using UI Path in multitenant_test.go.

**Comment by ShunjiTakano** [comment]
@friedrich-at-adobe Apologies for the late reply! (I didn't realise this PR got a reply 😅)
Yes let me dedicate some time to create test cases for this.

File: CHANGELOG.md (hunk 1 above)
Comment (ShunjiTakano) [reply CHANGELOG.md]: Good point, I have updated that to a bugfix

**Comment by alvinlin123** [comment]
I think this fix makes sense. I am not sure about the history on why the Alertmanager distributor only support limited path before; can't infer the reason from the commit history. But I think having the distributor to support all the path is right step forward.

**Review by alvinlin123 — APPROVED** [review]

**Review by friedrichg — COMMENTED** [review]
While I agree the commit fixes the problem. The PR does not add any tests to prevent the bug  from reappearing.

**Comment by alvinlin123** [comment]
Change in the PR was included in https://github.com/cortexproject/cortex/pull/5293/files. Closing this PR.

**Pull request closed without merging**